### PR TITLE
OAI-PMH refactor & ListMetadataFormats

### PR DESCRIPTION
**Pull request by thatbudakguy**
- Ensure earlestDatestamp displays with no items
- Refactor OaiController and extract OaiConcern
- Implement OAI-PMH ListMetadataFormats response

### CODE REVIEW

**Comment by thatbudakguy** [comment]
this doesn't address #711 but I figured we should either merge or change it before I start on that work.

File: app/controllers/oai_controller.rb [hunk 1]
```diff
@@ -170,8 +184,32 @@ def build_identify_response(earliest_date)
...
+
+  # See http://www.openarchives.org/OAI/openarchivesprotocol.html#ListMetadataFormats
+  def build_list_metadata_formats_response
+    Nokogiri::XML::Builder.new do |xml|
+      build_oai_response xml, list_metadata_formats_params do
+        xml.ListMetadataFormats do
+          xml.metadataFormat do
+            xml.metadataPrefix 'marc21'
+            xml.schema 'http://www.loc.gov/standards/marcxml/schema/MARC21slim.xsd'
+            xml.metadataNamespace 'http://www.loc.gov/MARC21/slim'
```
Comment (thatbudakguy) [inline app/controllers/oai_controller.rb:196]: i can't understand the difference between marc21 and marcxml, or whether this is right...

File: app/controllers/oai_controller.rb [hunk 2]
```diff
@@ -92,16 +100,29 @@ def list_record_normalized_dump_candidates
...
+      xml.request(params.to_hash) do
+        xml.text oai_url
+      end
+      yield xml
+    end
+  end
+
   # rubocop:disable Metrics/MethodLength
   # See https://www.openarchives.org/OAI/openarchivesprotocol.html#ListRecords
   def build_list_records_response(dump, next_dump = nil)
```
Comment (corylown) [inline app/controllers/oai_controller.rb:128]: The `build_request` method is now gone, but it's being called from line 137 in the `build_oai_response` method.

File: app/controllers/oai_controller.rb (hunk 1 above)
Comment (corylown) [reply app/controllers/oai_controller.rb:196]: I think what you have is correct. The terminology is confusing, but MARC21slim is an XML expression of MARC21.

**Review by corylown — CHANGES_REQUESTED** [review]
There's a method being called that's now missing -- see inline comments. Otherwise this looks good to me.

**Review by corylown — APPROVED** [review]

**Pull request merged**